R/mutate.R

```diff
@@ -0,0 +1,484 @@
+mutate_expr <- function(expr, vars_expr, data) {
+    if (is_survey(data)) {
+        expr <- coerce_tbl_svy(expr, data)
+        rlang::expr(!!expr %>% srvyr::mutate(!!!vars_expr))
+    } else {
+        rlang::expr(!!expr %>% dplyr::mutate(!!!vars_expr))
+    }
+}
+
+
+mutate_expr_i <- function(expr, vars_expr, data, ...) {
+    purrr::map(seq_along(vars_expr), function(i) {
+        vars_expr_i <- rlang::list2(
+            !!names(vars_expr)[i] := vars_expr[[i]],
+            !!!purrr::map(list(...), function(x) rlang::sym(x[i]))
+        )
+        expr <<- mutate_expr(expr, vars_expr_i, data)
+    })
+    expr
+}
+
+
+#' Combine variables into one categorical variable
+#'
+#' Combine chosen variables of any class by concatenating
+#' them into one factor variable, and returns the result
+#' along with tidyverse code used to generate it.
+#'
+#' @param data a dataframe with the columns to be combined
+#' @param vars  a character vector of the variables to be combined
+#' @param sep a character string to separate the levels
+#' @param name a name for the new variable
+#' @param keep_empty logical, if \code{FALSE} empty level combinations
+#'        are removed from the factor
+#' @param keep_na logical, if \code{TRUE} the \code{<NA>} in the factors or
+#'        \code{NA} in the characters will be replaced with \code{"(Missing)"};
+#'        otherwise, the resulting entries will return \code{<NA>}
+#'
+#' @return original dataframe containing new columns of the new
+#'         categorical variable with tidyverse code attached
+#' @rdname combine_vars
+#' @importFrom forcats fct_explicit_na
+#' @examples
+#' combined <- combine_vars(warpbreaks, vars = c("wool", "tension"), sep = "_")
+#' cat(code(combined))
+#' head(combined)
+#'
+#' @author Owen Jin, Stephen Su
+#' @export
+combine_vars <- function(data, vars, sep = ":", name = NULL,
```

Comment:
As per https://github.com/iNZightVIT/Lite/issues/281, can we add a check for the number of levels that will be created? If > 100, return an error.

Reply:
Is it for the purpose of saving computational time or avoid meaningless output with huge number of levels? That is, should I try to determine the number of levels before or after `forcats::fct_cross()` is called?

Reply:
Before, as it is supposedly causing some crashing. Can we also do something similar with `convert_to_cat()`? 

An option would be to use an option with default, so it can be overridden if needed.
e.g., 
```r
n_max <- getOption("inzighttools.max_levels", 100)
```

Reply:
I saw [iNZightVIT/Lite#281](https://github.com/iNZightVIT/Lite/issues/281) proposed a method of getting the number of factor levels of the new variable with the product of `length(unique(...))`, almost always overestimates the resulting number of factor levels. The most extreme example of the issue is `forcats::fct_cross(letters, letters)` which returns 26 levels instead of 26 * 26 levels. I am thinking of instead using `length(unique(paste(<all vars>)))` which returns accurately the resulting levels and is also much faster than calling `forcats::fct_cross()` as it does not involve actually combining the actual factor levels,

Reply:
This is only true if `keep_empty` is `FALSE` - if it's `TRUE`, then the result would be `N x M` (e.g., 26 x 26), so you'll need to handle that case.

Reply:
Could you please elaborate on the scenario more? I tried running `fct_explicit_na(letters) |> fct_cross(fct_explicit_na(letters)) |> nlevels()` the code returns `26`.

Reply:
```r
x <- factor(c("A", "B", "C"))
forcats::fct_cross(x, x, keep_empty = TRUE)
# [1] A:A B:B C:C
# Levels: A:A B:A C:A A:B B:B C:B A:C B:C C:C
```

Reply:
I see, I got `keep_na` and `keep_empty` mixed up.